File: core/src/commonMain/kotlin/com/bluzi/babymonitor/monitor/CameraControl.kt
package com.bluzi.babymonitor.monitor

import com.bluzi.babymonitor.data.AppStore
import com.bluzi.babymonitor.net.platformHttp
import com.bluzi.babymonitor.platform.ioDispatcher
import com.bluzi.babymonitor.xiaomi.MiCloud
import com.bluzi.babymonitor.xiaomi.NightVisionMode
import com.bluzi.babymonitor.xiaomi.XiaomiException
import kotlinx.coroutines.withContext

/**
 * LIVE-10 / PROTO-24: camera-side settings (night vision) over the MiOT cloud API. Separate from
 * the streaming connection — it builds a short-lived cloud client from the stored session, like
 * the device picker does. The mode lives on the camera, shared by all viewers.
 */
object CameraControl {
    // Camera Control service, night-shot property (C100 / chuangmi.camera.077ac1).
    private const val NIGHT_VISION_SIID = 2
    private const val NIGHT_VISION_PIID = 3

    private fun cloud(store: AppStore): Pair<MiCloud, String> {
        val session = store.loadSession() ?: throw XiaomiException("not signed in")
        val device = store.loadDevice() ?: throw XiaomiException("no camera selected")
        val cloud = MiCloud(platformHttp, session = session)
        cloud.onSessionRefreshed = { store.saveSession(it) } // AUTH-7
        return cloud to device.did
    }

    suspend fun getNightVision(store: AppStore): NightVisionMode? = withContext(ioDispatcher) {
        val (cloud, did) = cloud(store)
        val raw = cloud.miotGetProp(did, NIGHT_VISION_SIID, NIGHT_VISION_PIID)
        (raw as? Number)?.let { NightVisionMode.fromValue(it.toInt()) }
    }

    suspend fun setNightVision(store: AppStore, mode: NightVisionMode) = withContext(ioDispatcher) {
        val (cloud, did) = cloud(store)
        cloud.miotSetProp(did, NIGHT_VISION_SIID, NIGHT_VISION_PIID, mode.value)
    }
}

File: windows/src/BabyMonitor.Core/Monitor/CameraControl.cs
using BabyMonitor.Core.Data;
using BabyMonitor.Core.Net;
using BabyMonitor.Core.Xiaomi;

namespace BabyMonitor.Core.Monitor;

/// <summary>
/// LIVE-10 / PROTO-24: camera-side settings (night vision) over the MiOT cloud API. Separate from the
/// streaming connection — it builds a short-lived cloud client from the stored session, like the device
/// picker does. The mode lives on the camera, shared by all viewers.
/// </summary>
public static class CameraControl
{
    // Camera Control service, night-shot property (C100 / chuangmi.camera.077ac1).
    private const int NightVisionSiid = 2;
    private const int NightVisionPiid = 3;

    public static async Task<NightVisionMode?> GetNightVisionAsync(
        AppStore store,
        IMiHttp? http = null,
        CancellationToken ct = default)
    {
        var (cloud, did) = Cloud(store, http);
        var raw = await cloud.MiotGetPropAsync(did, NightVisionSiid, NightVisionPiid, ct).ConfigureAwait(false);
        return raw switch
        {
            long l => Mi.NightVisionFromValue((int)l),
            double d => Mi.NightVisionFromValue((int)d),
            _ => null,
        };
    }

    public static async Task SetNightVisionAsync(
        AppStore store,
        NightVisionMode mode,
        IMiHttp? http = null,
        CancellationToken ct = default)
    {
        var (cloud, did) = Cloud(store, http);
        await cloud.MiotSetPropAsync(did, NightVisionSiid, NightVisionPiid, (int)mode, ct).ConfigureAwait(false);
    }

    private static (MiCloud Cloud, string Did) Cloud(AppStore store, IMiHttp? http)
    {
        var session = store.LoadSession() ?? throw new XiaomiException("not signed in");
        var device = store.LoadDevice() ?? throw new XiaomiException("no camera selected");
        var cloud = new MiCloud(http ?? SystemMiHttp.Shared, session: session)
        {
            OnSessionRefreshed = s =>
            {
                store.SaveSession(s); // AUTH-7
                return Task.CompletedTask;
            },
        };
        return (cloud, device.Did);
    }
}
